iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,163 @@
                            const CaproIdString_t& service,
-                           const CaproIdString_t& instance) const
+                           const CaproIdString_t& instance) const noexcept
 {
-    if (instance == iox::cxx::string<100>(capro::AnyInstanceString))
+    cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> intersection;
+
+    // Find (K1, K2)
+    // O(log n + log n + max(#PossibleServices + #possiblesInstances) + #intersection)
+    if (instance != capro::IdString_t(capro::AnyInstanceString)
+        && service != capro::IdString_t(capro::AnyServiceString))
     {
-        for (auto& instance : m_serviceMap[service].instanceSet)
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleServices;
+        cxx::vector<uint64_t, MAX_SERVICE_DESCRIPTIONS> possibleInstances;
+
+        auto rangeServiceMap = m_serviceMap.equal_range(service);
+        for (auto entry = rangeServiceMap.first; entry != rangeServiceMap.second; ++entry)
+        {
+            possibleServices.push_back(entry->second);
+        }
+
+        auto rangeInstanceMap = m_instanceMap.equal_range(instance);
+        for (auto entry = rangeInstanceMap.first; entry != rangeInstanceMap.second; ++entry)
+        {
+            possibleInstances.push_back(entry->second);
+        }
+
+        ::std::set_intersection(possibleServices.begin(),
+                                possibleServices.end(),
+                                possibleInstances.begin(),
+                                possibleInstances.end(),
+                                ::std::back_inserter(intersection));
+
+        for (auto& value : intersection)
         {
-            instances.push_back(instance);
+            searchResult.push_back(m_serviceDescriptionVector[value]);
         }
     }
     else
     {
-        auto& instanceSet = m_serviceMap[service].instanceSet;
-        auto iter = std::find(instanceSet.begin(), instanceSet.end(), instance);
-        if (iter != instanceSet.end())
+        // Find (*, K2)
+        // O(log n + #result)
+        if (service == capro::IdString_t(capro::AnyServiceString)
+            && instance != capro::IdString_t(capro::AnyInstanceString))
+        {
+            auto range = m_instanceMap.equal_range(instance);
+            for (auto entry = range.first; entry != range.second; ++entry)
+            {
+                searchResult.push_back(m_serviceDescriptionVector[entry->second]);
+            }
+        }
+        // Find (K1, *)
+        // O(log n + #result)
+        else if (instance == capro::IdString_t(capro::AnyInstanceString)
+                 && service != capro::IdString_t(capro::AnyServiceString))
+        {
+            auto range = m_serviceMap.equal_range(service);
+            for (auto entry = range.first; entry != range.second; ++entry)
+            {
+                searchResult.push_back(m_serviceDescriptionVector[entry->second]);
+            }
```

Comment:
Since this is used quite often, can we have a lambda which could be used like
```cpp
storeIndices(containerToStore, range);
```

Reply:
With only two occurrences, I prefer to keep things how they are. `ServiceRegistry::find` is very readable and fits well on one monitor currently.

Reply:
well, I count 4 ;)

Reply:
any input on this? Resolving would be fine since the duplicated code is minimal, but you might revise your decision with four occurrences ;)

Reply:
any input on this?

Reply:
@elBoberido can you list your for occurrences here please so that @mossmaurice can see them as well.